src/main/java/org/opensearch/knn/index/query/iterators/VectorIdsKNNIterator.java

```diff
@@ -33,33 +31,44 @@ public class FilteredIdsKNNIterator implements KNNIterator {
     protected int docId;
     private final SegmentLevelQuantizationInfo segmentLevelQuantizationInfo;
 
-    FilteredIdsKNNIterator(
+    VectorIdsKNNIterator(
         final BitSet filterIdsBitSet,
         final float[] queryVector,
         final KNNFloatVectorValues knnFloatVectorValues,
         final SpaceType spaceType
-    ) {
+    ) throws IOException {
         this(filterIdsBitSet, queryVector, knnFloatVectorValues, spaceType, null, null);
     }
 
-    public FilteredIdsKNNIterator(
+    VectorIdsKNNIterator(final float[] queryVector, final KNNFloatVectorValues knnFloatVectorValues, final SpaceType spaceType)
+        throws IOException {
+        this(null, queryVector, knnFloatVectorValues, spaceType, null, null);
+    }
+
+    public VectorIdsKNNIterator(
         final BitSet filterIdsBitSet,
         final float[] queryVector,
         final KNNFloatVectorValues knnFloatVectorValues,
         final SpaceType spaceType,
         final byte[] quantizedQueryVector,
         final SegmentLevelQuantizationInfo segmentLevelQuantizationInfo
-    ) {
-        this.filterIdsBitSet = filterIdsBitSet;
-        this.bitSetIterator = new BitSetIterator(filterIdsBitSet, filterIdsBitSet.length());
+    ) throws IOException {
+        this.bitSetIterator = filterIdsBitSet == null ? null : new BitSetIterator(filterIdsBitSet, filterIdsBitSet.length());
         this.queryVector = queryVector;
         this.knnFloatVectorValues = knnFloatVectorValues;
         this.spaceType = spaceType;
-        this.docId = bitSetIterator.nextDoc();
+        this.docId = getNextDocId();
         this.quantizedQueryVector = quantizedQueryVector;
         this.segmentLevelQuantizationInfo = segmentLevelQuantizationInfo;
     }
 
+    protected int getNextDocId() throws IOException {
+        if (bitSetIterator != null) {
+            return this.bitSetIterator.nextDoc();
+        }
+        return this.knnFloatVectorValues.nextDoc();
```

Comment:
should we advance the vector values also in this function which will ensure that we don't need this code in nextDoc() function.

```
if (bitSetIterator != null) {
    knnFloatVectorValues.advance(docId);
}
```

Reply:
i thought about it, but I decided against that since advance will do nothing but pointing to itself in non filter cases. With this way i feel it is clear and conveys exactly why we are advancing during filter. However, do you see potential issue with impact in performance ?

Reply:
its not about impact in performance. if we want t advance to advance to next doc then we should actually do that for both bitsetIterator and also knnFloatVectorValues. Hence I believe either we should remove the call of nextDoc from constructors and make things clean or move this logic to getNextDocId(). I would prefer doing first one where we remove the call of nextDoc from constructor.

Reply:
I was planning towards remove that call from constructor but NestedIterator will break  if i move docId inside nextDoc. With regards to your suggestion, i believe it will look like below
```
    if (bitSetIterator == null) {
          return this.knnFloatVectorValues.nextDoc();
    }
    int docId = this.bitSetIterator.nextDoc();
   if(docId != DocIdSetIterator.NO_MORE_DOCS) {
          knnFloatVectorValues.advance(docId);
   }
   return docId
```
i was hesitant since i had to to add NO_MORE_DOCS check again.

Reply:
As discussed offline, adding comment to nextDoc method